### PR TITLE
Use requests.get to replace urllib.request.urlopen

### PR DESCRIPTION
**Pull request by sideshowbarker**
This change replaces usage of urllib in `client.py` with Requests.

Among other reasons for making the change, it helps prevent a *“CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate”* failure that can happen in macOS environments, as documented in the Bikeshed docs:

> Note: If you’re on a Mac, bikeshed update might fail with some errors about certificates. This is a known issue with the version of Python 3.7 shipped with XCode, which might be the default on your system. To fix this, install Python 3.7 with Brew instead.

See also https://github.com/tabatkins/bikeshed/issues/1352#issuecomment-610221695

### CODE REVIEW

File: json_home_client/client.py [hunk 1]
```diff
@@ -127,7 +127,7 @@ def content_type(self) -> MimeType:
     @property
     def encoding(self) -> str:
         """Get encoding of response."""
-        content_type = self.headers.get('content-type') if (self.headers) else None
+        content_type = self.headers['content-type'] if (self.headers) else None
```
Comment (tabatkins) [inline json_home_client/client.py]: This should likely stay as it is; the old code gracefully returned None if 'content-type' wasn't in the headers, but the new code will panic.

**Review by tabatkins — CHANGES_REQUESTED** [review]

**Comment by plinss** [comment]
So it's not an issue for Bikeshed since Shepherd uses a real cert, but I've had problems with requests not honoring custom CA roots. So I'm reluctant to switch over for a general purpose library.

I have some code from another project that hooks urllib.request's SSL negotiation, I'll see if that addresses the XCode issue.

File: json_home_client/client.py (hunk 1 above)
Comment (sideshowbarker) [reply json_home_client/client.py]: OK, thanks — reverted that change

**Comment by sideshowbarker** [comment]
> I've had problems with requests not honoring custom CA roots

I’m curious about what specific problems. I take it that just setting `REQUESTS_CA_BUNDLE` in the environment isn’t enough to actually make it work? (or just passing it with a `verify` param)

https://incognitjoe.github.io/adding-certs-to-requests.html looks like an interesting way of dynamically updating a CA bundle when an SSL error occurs.

**Comment by plinss** [comment]
Setting environment variables isn't always an option, and having to pass the path to the ca bundle can also be problematic for cross-platform code. It's far easier to simply rely on the system's default ca bundle. That way those using custom CAs can just drop their CA cert in one place and be done with it.

I also can't reproduce this error locally.

**Comment by sideshowbarker** [comment]
> I also can't reproduce this error locally.

I can reproduce it in my environment with `/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.7/bin/python3` and doing:

```
bikeshed update --skip-manifest --test-suites
```
…to make sure that it’s hitting one of the json_home_client `APIClient` calls that Bikeshed makes.

With that I get this error:

```
Downloading test suite data...
FATAL ERROR: Couldn't download test suite data.  Error was:
Network error: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1056)
 ✘  Did not generate, due to fatal errors
```

Here’s the most-relevant traceback:
```
Traceback (most recent call last):
  File "bikeshed.py", line 5, in <module>
    bikeshed.main()
  File "/Users/mike/workspace/bikeshed/bikeshed/cli.py", line 210, in main
    handleUpdate(options, extras)
  File "/Users/mike/workspace/bikeshed/bikeshed/cli.py", line 238, in handleUpdate
    update.update(anchors=options.anchors, backrefs=options.backrefs, biblio=options.biblio, caniuse=options.caniuse, mdn=options.mdn, linkDefaults=options.linkDefaults, testSuites=options.testSuites, languages=options.languages, wpt=options.wpt, dryRun=constants.dryRun, force=options.force)
  File "/Users/mike/workspace/bikeshed/bikeshed/update/main.py", line 42, in update
    "testSuites": updateTestSuites.update(path=path, dryRun=dryRun) if testSuites else None,
  File "/Users/mike/workspace/bikeshed/bikeshed/update/updateTestSuites.py", line 16, in update
    shepherd = APIClient("https://api.csswg.org/shepherd/", version="vnd.csswg.shepherd.v1")
  File "/Library/Python/3.7/site-packages/json_home_client-0.0.0-py3.7.egg/json_home_client/client.py", line 247, in __init__
  File "/Library/Python/3.7/site-packages/json_home_client-0.0.0-py3.7.egg/json_home_client/client.py", line 269, in _load_home
  File "/Library/Python/3.7/site-packages/json_home_client-0.0.0-py3.7.egg/json_home_client/client.py", line 326, in _request
json_home_client.client.NetworkError: Network error: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1056)
```

…and here’s the full set of tracebacks:

```
Traceback (most recent call last):
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 1317, in do_open
    encode_chunked=req.has_header('Transfer-encoding'))
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/http/client.py", line 1229, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/http/client.py", line 1275, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/http/client.py", line 1224, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/http/client.py", line 1016, in _send_output
    self.send(msg)
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/http/client.py", line 956, in send
    self.connect()
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/http/client.py", line 1392, in connect
    server_hostname=server_hostname)
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/ssl.py", line 412, in wrap_socket
    session=session
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/ssl.py", line 853, in _create
    self.do_handshake()
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/ssl.py", line 1117, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1056)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Library/Python/3.7/site-packages/json_home_client-0.0.0-py3.7.egg/json_home_client/client.py", line 321, in _request
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 222, in urlopen
    return opener.open(url, data, timeout)
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 525, in open
    response = self._open(req, data)
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 543, in _open
    '_open', req)
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 503, in _call_chain
    result = func(*args)
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 1360, in https_open
    context=self._context, check_hostname=self._check_hostname)
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/urllib/request.py", line 1319, in do_open
    raise URLError(err)
urllib.error.URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1056)>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "bikeshed.py", line 5, in <module>
    bikeshed.main()
  File "/Users/mike/workspace/bikeshed/bikeshed/cli.py", line 210, in main
    handleUpdate(options, extras)
  File "/Users/mike/workspace/bikeshed/bikeshed/cli.py", line 238, in handleUpdate
    update.update(anchors=options.anchors, backrefs=options.backrefs, biblio=options.biblio, caniuse=options.caniuse, mdn=options.mdn, linkDefaults=options.linkDefaults, testSuites=options.testSuites, languages=options.languages, wpt=options.wpt, dryRun=constants.dryRun, force=options.force)
  File "/Users/mike/workspace/bikeshed/bikeshed/update/main.py", line 42, in update
    "testSuites": updateTestSuites.update(path=path, dryRun=dryRun) if testSuites else None,
  File "/Users/mike/workspace/bikeshed/bikeshed/update/updateTestSuites.py", line 16, in update
    shepherd = APIClient("https://api.csswg.org/shepherd/", version="vnd.csswg.shepherd.v1")
  File "/Library/Python/3.7/site-packages/json_home_client-0.0.0-py3.7.egg/json_home_client/client.py", line 247, in __init__
  File "/Library/Python/3.7/site-packages/json_home_client-0.0.0-py3.7.egg/json_home_client/client.py", line 269, in _load_home
  File "/Library/Python/3.7/site-packages/json_home_client-0.0.0-py3.7.egg/json_home_client/client.py", line 326, in _request
json_home_client.client.NetworkError: Network error: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1056)
```

**Comment by sideshowbarker** [comment]
> Setting environment variables isn't always an option

I’m not disagreeing, but I’m curious when it’s not an option. I mean, I wonder under what circumstances putting `os.environ['REQUESTS_CA_BUNDLE']` in code wouldn’t be an option.

> and having to pass the path to the ca bundle can also be problematic for cross-platform code.

Yeah, that part I can imagine being problematic

> It's far easier to simply rely on the system's default ca bundle.

The specific problem with the Xcode-installed Python is that it ignores the system’s default ca bundle. It’s intentional on the part of Apple, and documented in some bug reports. I’ll dig up the links and post them here or in https://github.com/tabatkins/bikeshed/issues/1352

**Comment by plinss** [comment]
> The specific problem with the Xcode-installed Python is that it ignores the system’s default ca bundle. It’s intentional on the part of Apple, and documented in some bug reports.

Ok, digging around apparently installing the 'certifi' package fixes that (at least for some versions). Can you try that locally? I'd rather just add that as a macOS dependency than switch to requests.

**Comment by sideshowbarker** [comment]
> Ok, digging around apparently installing the 'certifi' package fixes that (at least for some versions). Can you try that locally?

Yup, will try it right now

**Comment by sideshowbarker** [comment]
> Ok, digging around apparently installing the 'certifi' package fixes that (at least for some versions). Can you try that locally?

I tried it and still get exactly the same failure.

I think for it to have any effect in the context of this issue, any `urllib.request.urlopen` call would also need to be changed to add the `cafile=certifi.where()` param and value:

```
urllib.request.Request(url=url, data=payload, method=method, cafile=certifi.where()
```

**Comment by plinss** [comment]
Closing this since the new code can accept a custom ca bundle. To get the same behavior as requests call the client constructor with `ca_cert_path=certifi.where()`, otherwise it will use the system default bundle and pick up custom ca certs installed there (which requests will not do).

Thanks for pointing out the issue tho and all the help debugging.

**Pull request closed without merging**